che-incubator/kubernetes-image-puller-operator
# Add validation webhook

Related issue: https://issues.redhat.com/browse/CRW-4980

This PR creates a validation webhook that validates the CREATE event for KubernetesImagePuller custom resources.
The validation webhook ensures that only users/SAs with the daemonset create permissions can create KubernetesImagePuller custom resources.

## How to test this PR
1. Build the image (optional)
I've built it here: `quay.io/dkwon17/kubernetes-image-puller-operator:validation-webhook` already.

The image can be built by running this command:
```
make docker-build IMG=quay.io/<username>/kubernetes-image-puller-operator:<tag>
```

2. In a new cluster, deploy the operator
```
make deploy IMG=quay.io/dkwon17/kubernetes-image-puller-operator:validation-webhook
```

3. Create two service accounts and necessary roles by running this command:
```
curl https://gist.githubusercontent.com/dkwon17/dfaed5ccfcb6ac5a414f578e105fdf4e/raw/6dd8a17e1614d8af689c7dc8495ed313857c6d8a/sa-template.yaml | oc process -f - -p NAMESPACE=kubernetes-image-puller-operator  | oc apply -f -
```

To summarize, these service accounts are created:
`no-create-daemonset` - does not have daemonset create permissions
`create-daemonset` - has daemonset create permissions


4. Try to create a KIP resource with the `no-create-daemonset` SA
```
kubectl apply -f - --as=system:serviceaccount:kubernetes-image-puller-operator:no-create-daemonset<<EOF
apiVersion: che.eclipse.org/v1alpha1
kind: KubernetesImagePuller
metadata:
  name: example
  namespace: kubernetes-image-puller-operator
spec: {}
EOF
```

You should get the following error:
```
Error from server (User "system:serviceaccount:kubernetes-image-puller-operator:no-create-daemonset"
not allowed to create daemonsets): error when creating "STDIN": admission webhook "vkubernetesimagepuller.kb.io" 
denied the request: User "system:serviceaccount:kubernetes-image-puller-operator:no-create-daemonset" not
allowed to create daemonsets
```

5. Try to create a KIP resource with the `create-daemonset` SA:
 ```
kubectl apply -f - --as=system:serviceaccount:kubernetes-image-puller-operator:create-daemonset<<EOF
apiVersion: che.eclipse.org/v1alpha1
kind: KubernetesImagePuller
metadata:
  name: example
  namespace: kubernetes-image-puller-operator
spec: {}
EOF
```

It should work:
```
kubernetesimagepuller.che.eclipse.org/example created
```

6. Run `make test` to run all tests.
